test/e2e-go/features/transactions/messageRate_test.go

```diff
@@ -0,0 +1,179 @@
+		sleepTime := time.Duration(int64(math.Round(sleepDuration*1000))) * time.Millisecond
+		time.Sleep(sleepTime)
+	}
+}
+
+func testMessageRateChangesWithTxnRate(t *testing.T, templatePath string, txnRates []uint) {
+	t.Parallel()
+	a := require.New(fixtures.SynchronizedTest(t))
+
+	var fixture fixtures.RestClientFixture
+	fixture.SetupNoStart(t, templatePath)
+	configDir := fixture.PrimaryDataDir()
+	cfg, err := config.LoadConfigFromDisk(configDir)
+	a.NoError(err)
+	cfg.EnableVerbosedTransactionSyncLogging = true
+	cfg.SaveToDisk(configDir)
+	fixture.Start()
+
+	defer fixture.Shutdown()
+
+	pingClient := fixture.LibGoalClient
+	pingAccountList, err := fixture.GetWalletsSortedByBalance()
+	a.NoError(err, "fixture should be able to get wallets sorted by balance")
+	pingAccount := pingAccountList[0].Address
+
+	pongClient := fixture.GetLibGoalClientForNamedNode("Node")
+	pongAccounts, err := fixture.GetNodeWalletsSortedByBalance(pongClient.DataDir())
+	a.NoError(err)
+	var pongAccount string
+	for _, acct := range pongAccounts {
+		if acct.Address == pingAccount {
+			continue
+		}
+		// we found an account.
+		pongAccount = acct.Address
+		break
+	}
+
+	a.NotEqual(pingAccount, pongAccount, "accounts under study should be different")
+
+	minTxnFee, minAcctBalance, err := fixture.CurrentMinFeeAndBalance()
+	a.NoError(err)
+
+	transactionFee := minTxnFee + 5
+	amountPingSendsPong := minAcctBalance * 3 / 2
+
+	// build the path for the primary node's log file
+	var logPath strings.Builder
+	logPath.WriteString(fixture.PrimaryDataDir())
+	logPath.WriteString("/node.log")
+
+	// seek to `bytesRead` bytes when reading the log file
+	bytesRead := 0
+
+	// store message rate for each of the txn rates
+	prevMsgRate := 0.0
+
+	for _, txnRate := range txnRates {
+		startTime := time.Now()
+		txnSentCount := uint(0)
+
+		for {
+			// send txns at txnRate for 30s
+			timeSinceStart := time.Since(startTime)
+			if timeSinceStart > 30*time.Second {
+				break
+			}
+
+			_, err := pingClient.SendPaymentFromUnencryptedWallet(pingAccount, pongAccount, transactionFee, amountPingSendsPong, GenerateRandomBytes(8))
+			a.NoError(err, "fixture should be able to send money (ping -> pong)")
+			txnSentCount++
+
+			throttleTransactionRate(startTime, txnRate, txnSentCount)
+		}
+
+		// parse the log to find out message rate and bytes of the log file read
+		msgRate, newBytesRead, err := parseLog(logPath.String(), bytesRead)
+		a.NoError(err)
+		aErrorMessage := fmt.Sprintf("TxSync message rate not monotonic for txn rate: %d", txnRate)
+		a.GreaterOrEqual(msgRate, prevMsgRate, aErrorMessage)
+		bytesRead += newBytesRead
+		prevMsgRate = msgRate
+	}
+
+}
+
+func parseLog(logPath string, startByte int) (float64, int, error) {
+	file, err := os.Open(logPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer file.Close()
+
+	// start reading log file from startByte
+	_, err = file.Seek(int64(startByte), 0)
```

Comment:
I think something is going wrong here. 
Can you verify with the sequence number in  "Outgoing Txsync \#seqNum" that each time you are starting from the right line in the log?
If this is wrong and is getting tricky, you can use the sequence number to anchor where to start reading from the file.

Reply:
Fixed.